### PR TITLE
Refactor strategy configuration and add a primitive hook for known good reduction configurations.

### PR DESCRIPTION
**Pull request by nicolasvasilache**
Improve doc for matchAndSetReductionStrategy.
Refactor strategy configuration and introduce a hardware model shim.
Add a primitive hook for known good reduction configurations.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/TransformDialectStrategies/CPU/Common.h [hunk 1]
```diff
@@ -20,12 +20,24 @@ namespace cpu {
...
+/// Map an N-D parallel, 1-D reduction operation with optional leading and
+/// optional trailing elementwise operations.
+/// The 1-D reduction dimension must be in the most minor dimension.
+/// The innermost dimensions of the leading and trailing operations must be most
+/// minor along all accesses.
+/// Return failure if matching fails.
+/// On a successful match, configure a reduction strategy based on a proxy model
+/// of the hardware and construct transform dialect IR that implements the
+/// reduction strategy. The transform dialect IR is added using
+/// `createTransformRegion`, which creates a top-level ModuleOp after the
```
Comment (ftynse) [inline compiler/src/iree/compiler/Codegen/TransformDialectStrategies/CPU/Common.h:50]: Nit: there's no `createTransformRegion` visible here.

File: compiler/src/iree/compiler/Codegen/TransformDialectStrategies/GPU/Common.h [hunk 2]
```diff
@@ -80,9 +81,17 @@ void build1DSplittingStrategyWithOptionalThreadMapping(
...
+/// Map an N-D parallel, 1-D reduction operation with optional leading and
+/// optional trailing elementwise operations.
+/// The 1-D reduction dimension must be in the most minor dimension.
+/// The innermost dimensions of the leading and trailing operations must be most
+/// minor along all accesses.
+/// Return failure if matching fails.
+/// On a successful match, configure a reduction strategy based on a proxy model
+/// of the hardware and construct transform dialect IR that implements the
+/// reduction strategy. The transform dialect IR is added using
+/// `createTransformRegion`, which creates a top-level ModuleOp after the
```
Comment (ftynse) [inline compiler/src/iree/compiler/Codegen/TransformDialectStrategies/GPU/Common.h:102]: Ditto.

File: compiler/src/iree/compiler/Codegen/TransformDialectStrategies/CPU/Common.cpp [hunk 3]
```diff
@@ -76,6 +76,22 @@ std::pair<Value, Value> mlir::iree_compiler::cpu::buildCommonTrailingStrategy(
...
+/// precedence over other heuristics. In the future, this could be lifted to
+/// e.g. `cpuModel` or higher up in some transform dialect database summary of
+/// "known good things".
+static FailureOr<ReductionConfig> applyKnownGoodReductionConfigurations(
+    const transform_ext::MatchedReductionCaptures &captures,
+    const CPUModel &cpuModel) {
+  int64_t reductionSize = captures.reductionOpSizes.back();
+  if (cpuModel.model == CPUModel::kXeonGold6154) {
+    if (captures.reductionOutputElementalTypeBitWidth == 32) {
+      if (reductionSize) return ReductionConfig{32};
```
Comment (ftynse) [inline compiler/src/iree/compiler/Codegen/TransformDialectStrategies/CPU/Common.cpp:89]: Nit: this config looks too much like magic numbers. Can we at least have `/*foo=*/` comments in the constructor?

File: compiler/src/iree/compiler/Codegen/TransformDialectStrategies/CPU/Common.h [hunk 4]
```diff
@@ -34,7 +34,8 @@ std::pair<Value, Value> buildCommonTrailingStrategy(ImplicitLocOpBuilder& b,
 /// to configure the reduction strategy. In the future, this will need to be
 /// driven by some contract with the runtime.
 struct CPUModel {
-  StringRef model = "Intel-Xeon-Gold-6154-3GHz";
+  static constexpr StringRef kXeonGold6154 = "Intel-Xeon-Gold-6154-3GHz";
```
Comment (ftynse) [inline compiler/src/iree/compiler/Codegen/TransformDialectStrategies/CPU/Common.h:37]: Nit: StringLiteral

File: compiler/src/iree/compiler/Codegen/TransformDialectStrategies/Common/AbstractReductionStrategy.h [hunk 5]
```diff
@@ -34,8 +34,6 @@ struct AbstractReductionStrategy {
 
   virtual ~AbstractReductionStrategy() {}
```
Comment (ftynse) [inline compiler/src/iree/compiler/Codegen/TransformDialectStrategies/Common/AbstractReductionStrategy.h:35]: It looks like there's nothing virtual left so we can devirtualize the destructor.

File: compiler/src/iree/compiler/Codegen/TransformDialectStrategies/GPU/Common.h [hunk 6]
```diff
@@ -7,6 +7,8 @@
 #ifndef IREE_COMPILER_CODEGEN_TRANSFORM_DIALECT_STRATEGIES_GPU_COMMON_H_
 #define IREE_COMPILER_CODEGEN_TRANSFORM_DIALECT_STRATEGIES_GPU_COMMON_H_
 
+#include <iree/compiler/Codegen/TransformDialectStrategies/GPU/AbstractReductionStrategy.h>
```
Comment (ftynse) [inline compiler/src/iree/compiler/Codegen/TransformDialectStrategies/GPU/Common.h:10]: Nit: I suppose you want `"` here

**Review by ftynse — COMMENTED** [review]
Do I read it correctly that the transform dialect-based strategy will only kick in for a very specific target CPU/GPU? It's okay for the default setting, but we need a mechanism to override that and have it kick in on _any_ hardware so we can, at least, measure the performance without recompiling the compiler.

**Comment by stellaraccident** [comment]
> Do I read it correctly that the transform dialect-based strategy will only kick in for a very specific target CPU/GPU? It's okay for the default setting, but we need a mechanism to override that and have it kick in on _any_ hardware so we can, at least, measure the performance without recompiling the compiler.

I had the same question and read this as needing a mechanism that doesn't exist but wanting to at least now that it bifurcates here.

**Comment by nicolasvasilache** [comment]
> > Do I read it correctly that the transform dialect-based strategy will only kick in for a very specific target CPU/GPU? It's okay for the default setting, but we need a mechanism to override that and have it kick in on _any_ hardware so we can, at least, measure the performance without recompiling the compiler.
> 
> I had the same question and read this as needing a mechanism that doesn't exist but wanting to at least now that it bifurcates here.

Nope, it always kicks in and is just a placeholder, I'll rename to default or something.

**Comment by ftynse** [comment]
> I had the same question and read this as needing a mechanism that doesn't exist but wanting to at least now that it bifurcates here.

Agreed. Some sort of selector over a combination of target HW and workload that results in one of ("on in production", "staging, tested but not currently on", "untested experimental") with potential extension.

**Review by ftynse — APPROVED** [review]
It would be nice to eventually connect all configurations in a single dispatch.

File: compiler/src/iree/compiler/Codegen/TransformDialectStrategies/GPU/Common.cpp [hunk 7]
```diff
@@ -243,11 +246,104 @@ std::pair<Value, Value> mlir::iree_compiler::gpu::buildCommonTrailingStrategy(
...
+static FailureOr<ReductionConfig> applyKnownGoodReductionConfigurations(
+    const transform_ext::MatchedReductionCaptures &captures,
+    const GPUModel &gpuModel) {
+  auto staged = ReductionStrategy::Staged;
+  int64_t reductionSize = captures.reductionOpSizes.back();
+  if (gpuModel.model == GPUModel::kDefaultGPU) {
+    if (captures.reductionOutputElementalTypeBitWidth == 32) {
+      if (reductionSize == 64) return ReductionConfig{64, 1, staged};
+      if (reductionSize == 128) return ReductionConfig{32, 4, staged};
+      if (reductionSize == 512) return ReductionConfig{256, 2, staged};
```
[the remark this answers is not in the document]
Comment (nicolasvasilache) [reply compiler/src/iree/compiler/Codegen/TransformDialectStrategies/GPU/Common.cpp:263]: `vector size 2 for size 512 is better` -> empirically yes, locally. This is where we'll want to scale up and save.

[the remark this answers is not in the document]
Comment (nicolasvasilache) [reply compiler/src/iree/compiler/Codegen/TransformDialectStrategies/GPU/Common.cpp:263]: `Can we add a /*numThreads=*/64-style comment?` -> #11778 11778

**Pull request merged**